File: single_number/single_number.rb
class Solution
  def single_number(nums)
    nums.reduce(:^) # XOR approach
  end
end


RSpec.describe Solution do
  it 'finds the single number' do
    solution = Solution.new
    expect(solution.single_number([2,2,1])).to eq(1)
    expect(solution.single_number([4,1,2,1,2])).to eq(4)
    expect(solution.single_number([1])).to eq(1)
  end
end

File: single_number/single_number.exs
defmodule Solution do
  def single_number(nums) do
    Enum.reduce(nums, 0, &Bitwise.bxor/2) # XOR approach
  end
end

defmodule SolutionTest do
  use ExUnit.Case
  doctest Solution

  test "finds the single number" do
    assert Solution.single_number([2, 2, 1]) == 1
    assert Solution.single_number([4, 1, 2, 1, 2]) == 4
    assert Solution.single_number([1]) == 1
  end
end
